### PR TITLE
Adding range support to PyPrimitive

### LINKED ISSUE #5391: Add support for ranges
## Description
We don't have support for `range` objects. We would like to add ranges in our primitive supported types.

Definition of DONE:
- [ ] Add some serde tests
- [ ] Make sure Ranges works on other types like Tensor or Pandas
- [ ] Add to `primitive_factory`

## Additional Context
We would like to have wrappers for range objects, to be able to manipulate them.

### PR DESCRIPTION
**Pull request by ArtistBanda**
## Description
fixes #5391 
Adds support for `range` object as `PyPrimitive`

## Affected Dependencies
None

## How has this been tested?
Added serde tests and integrated cpython range tests

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

**Comment by vsquareg** [comment]
Hey @ArtistBanda!

I did some work on this earlier (but never got to create a PR): [vsquareg/range-support1 ](https://github.com/vsquareg/PySyft/tree/range-support1). Range support is added, just the tests are pending.

You can take a look, if that's useful.

File: proto/lib/python/range.proto [hunk 1]
```diff
@@ -0,0 +1,14 @@
+syntax = "proto3";
+
+package syft.lib.python;
+import "proto/core/common/common_object.proto";
+
+message Range {
+  int64 start = 1;
+  int64 stop = 2;
+  int64 step = 3;
+  bool has_start = 4;
```
Comment (tudorcebere) [inline proto/lib/python/range.proto:10]: I think this has been introduced in Slices, but it's not necessary at all. You can validate if a field is Set with the `HasField` method. Could you address this and make the change in slices as well?

File: src/syft/lib/python/range.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+# syft relative
+from ...core.common import UID
+from ...proto.lib.python.range_pb2 import Range as Range_PB
+from .primitive_factory import PyPrimitive
+
+
+class Range(PyPrimitive):
+    def __init__(
+        self,
+        start: Any = None,
```
Comment (tudorcebere) [inline src/syft/lib/python/range.py:15]: Shouldn't this be int/Int?

File: src/syft/lib/python/range.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        stop: Union[Any] = None,
```
Comment (tudorcebere) [inline src/syft/lib/python/range.py:16]: Same as above

File: src/syft/lib/python/range.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        step: Union[Any] = None,
```
Comment (tudorcebere) [inline src/syft/lib/python/range.py:17]: Same as above

**Review by tudorcebere — COMMENTED** [review]
Nice job so far! Added a few comments :)

File: proto/lib/python/range.proto (hunk 1 above)
Comment (ArtistBanda) [reply proto/lib/python/range.proto:10]: Sure, will do.

File: src/syft/lib/python/range.py (hunk 2 above)
Comment (ArtistBanda) [reply src/syft/lib/python/range.py:15]: `mypy` had an issue with `int`, with `Any` it seems to be fine.

**Comment by ArtistBanda** [comment]
> Hey @ArtistBanda!
> 
> I did some work on this earlier (but never got to create a PR): [vsquareg/range-support1 ](https://github.com/vsquareg/PySyft/tree/range-support1). Range support is added, just the tests are pending.
> 
> You can take a look, if that's useful.

This looks great, you can add changes to this PR if you want.

**Comment by vsquareg** [comment]
> > Hey @ArtistBanda!
> > I did some work on this earlier (but never got to create a PR): [vsquareg/range-support1 ](https://github.com/vsquareg/PySyft/tree/range-support1). Range support is added, just the tests are pending.
> > You can take a look, if that's useful.
> 
> This looks great, you can add changes to this PR if you want.

I'm not sure I have the permission to do that. If you want to incorporate those changes, you can merge the branch into yours.

**Comment by ArtistBanda** [comment]
@vsquareg merged your branch, will pick up work from there. Thank you!

File: src/syft/lib/python/range.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        res = self.value.__ne__(other)
+        return PrimitiveFactory.generate_primitive(value=res)
+
+    def __sizeof__(self) -> SyPrimitiveRet:
+        res = self.value.__sizeof__()
+        return PrimitiveFactory.generate_primitive(value=res)
+
+    def __bool__(self) -> SyPrimitiveRet:
+        res = self.value.__bool__()
+        return PrimitiveFactory.generate_primitive(value=res)
```
Comment (vsquareg) [inline src/syft/lib/python/range.py:66]: This should return __bool__ object, not pointer (as in https://github.com/OpenMined/PySyft/pull/5452/commits/5ce1dd63b5246be7ec94f38af91e8815f8906ac6).

Comment (ArtistBanda) [reply src/syft/lib/python/range.py:66]:
```python
from syft.lib.python.range import Range
print(type(Range(2).__bool__()))
```
this prints 
`<class 'syft.lib.python.Bool'>`

So, this is returning a PySyft bool object.

**Comment by ArtistBanda** [comment]
`black` and `flake8` are in conflict here 
```python
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

tests/syft/lib/python/range/range_test.py:201:31: E203 whitespace before ':'
        self.assertEqual(x[idx : idx + 1][0], a + idx)
                              ^
tests/syft/lib/python/range/range_test.py:219:31: E203 whitespace before ':'
        self.assertEqual(x[idx : idx + 1][0], a + idx)
                              ^
tests/syft/lib/python/range/range_test.py:238:31: E203 whitespace before ':'
        self.assertEqual(x[idx : idx + 1][0], a + (idx * c))
                              ^
tests/syft/lib/python/range/range_test.py:257:31: E203 whitespace before ':'
        self.assertEqual(x[idx : idx + 1][0], a + (idx * c))
                              ^
4     E203 whitespace before ':'
4
```

If this is fixed, it is changed back by `black`

**Comment by ArtistBanda** [comment]
I have tested some of the CPython tests after serialization and deserialization to check the validity of the code and it works fine. @tudorcebere.

**Comment by ArtistBanda** [comment]
Please also let me know how to handle the `flake8` and `black` conflict.

**Comment by Param-29** [comment]
@ArtistBanda for your flake8 and black conflict, Id suggest adding `# noqa: E203` as I think there is nothing that we can do about this formatting conflict. 
Documentation: [Ignoring flake8 errors](https://flake8.pycqa.org/en/3.1.1/user/ignoring-errors.html)

A couple of things: 
1. If you have not noticed, your mypy is also failing 
```
src/syft/lib/python/range.py:65: error: "range" has no attribute "__bool__"
Found 1 error in 1 file (checked 379 source files)
```

2. Please re-base your branch as it makes reviewing a lot simpler. (Note: After this, you will also have to make changes in the code such as adding documentation based on newer changes we have added in pre-commit)

I hope this helps!

**Comment by ArtistBanda** [comment]
> @ArtistBanda for your flake8 and black conflict, Id suggest adding `# noqa: E203` as I think there is nothing that we can do about this formatting conflict.
> Documentation: [Ignoring flake8 errors](https://flake8.pycqa.org/en/3.1.1/user/ignoring-errors.html)
> 
> A couple of things:
> 
> 1. If you have not noticed, your mypy is also failing
> 
> ```
> src/syft/lib/python/range.py:65: error: "range" has no attribute "__bool__"
> Found 1 error in 1 file (checked 379 source files)
> ```
> 
> 1. Please re-base your branch as it makes reviewing a lot simpler. (Note: After this, you will also have to make changes in the code such as adding documentation based on newer changes we have added in pre-commit)
> 
> I hope this helps!

I noticed the `mypy` failing, I am a bit stuck with that, could you help me?

**Review by tudorcebere — APPROVED** [review]

**Pull request merged**